objectivec/ort_checkpoint.mm

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import "ort_checkpoint_internal.h"
+
+#include <optional>
+#include <string>
+#include <variant>
+#import "cxx_api.h"
+
+#import "error_utils.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@implementation ORTCheckpoint {
+  std::optional<Ort::CheckpointState> _checkpoint;
+}
+
+- (nullable instancetype)initWithCXXAPIOrtCheckPointFromPath:(NSString*)path
+                                                       error:(NSError**)error {
+  if ((self = [super init]) == nil) {
+    return nil;
+  }
+
+  try {
+    _checkpoint = Ort::CheckpointState::LoadCheckpoint(path.UTF8String);
+    return self;
+  }
+  ORT_OBJC_API_IMPL_CATCH_RETURNING_NULLABLE(error)
+}
+
++ (nullable instancetype)loadCheckpointFromPath:(NSString*)path
+                                          error:(NSError**)error {
+  return [[ORTCheckpoint alloc] initWithCXXAPIOrtCheckPointFromPath:path
+                                                              error:error];
+}
+
+- (BOOL)saveCheckpointToPath:(NSString*)path
+          withOptimizerState:(BOOL)includeOptimizerState
+                       error:(NSError**)error {
+  try {
+    Ort::CheckpointState::SaveCheckpoint([self CXXAPIOrtCheckpoint], path.UTF8String, includeOptimizerState);
+    return YES;
+  }
+  ORT_OBJC_API_IMPL_CATCH_RETURNING_BOOL(error)
+}
+
+- (BOOL)addPropertyWithName:(NSString*)name
+                   intValue:(int64_t)value
+                      error:(NSError**)error {
+  try {
+    [self CXXAPIOrtCheckpoint].AddProperty(name.UTF8String, value);
+    return YES;
+  }
+  ORT_OBJC_API_IMPL_CATCH_RETURNING_BOOL(error)
+}
+
+- (BOOL)addPropertyWithName:(NSString*)name
+                 floatValue:(float)value
+                      error:(NSError**)error {
+  try {
+    [self CXXAPIOrtCheckpoint].AddProperty(name.UTF8String, value);
+    return YES;
+  }
+  ORT_OBJC_API_IMPL_CATCH_RETURNING_BOOL(error)
+}
+
+- (BOOL)addPropertyWithName:(NSString*)name
+                stringValue:(NSString*)value
+                      error:(NSError**)error {
+  try {
+    [self CXXAPIOrtCheckpoint].AddProperty(name.UTF8String, value.UTF8String);
+    return YES;
+  }
+  ORT_OBJC_API_IMPL_CATCH_RETURNING_BOOL(error)
+}
+
+- (nullable NSString*)getStringPropertyWithName:(NSString*)name error:(NSError**)error {
+  Ort::Property value;
```

Comment:
consider putting as much of the implementation as possible into the `try { }`. now, it looks fine, but in the future if we add something that may throw it may be uncaught.

the helper macro ORT_CXX_API_THROW might be useful for handling errors.

https://github.com/microsoft/onnxruntime/blob/abd94b65b79bbca72e0640664018bd629683dde3/include/onnxruntime/core/session/onnxruntime_cxx_api.h#L75